lib/openapi.js

```diff
@@ -0,0 +1,262 @@
+'use strict'
+
+const fs = require('fs')
+const path = require('path')
+const yaml = require('js-yaml')
+const { formatParamUrl, plainJsonObjectToSwagger2, swagger2ParametersToOpenapi3, stripBasePathByServers } = require('./util')
+
+module.exports = function (opts, routes, Ref, cache, next) {
+  let ref
+
+  const info = opts.openapi.info || null
+  const servers = opts.openapi.servers || null
+  const components = opts.openapi.components || null
+  const tags = opts.openapi.tags || null
+  const externalDocs = opts.openapi.externalDocs || null
+
+  const stripBasePath = opts.stripBasePath
+  const transform = opts.transform
+  const hiddenTag = opts.hiddenTag
+  const extensions = []
+
+  for (const [key, value] of Object.entries(opts.openapi)) {
+    if (key.startsWith('x-')) {
+      extensions.push([key, value])
+    }
+  }
+
+  return function (opts) {
+    if (opts && opts.yaml) {
+      if (cache.swaggerString) return cache.swaggerString
+    } else {
+      if (cache.swaggerObject) return cache.swaggerObject
+    }
+
+    const swaggerObject = {}
+    let pkg
+
+    try {
+      pkg = JSON.parse(fs.readFileSync(path.join(__dirname, '..', 'package.json')))
+    } catch (err) {
+      return next(err)
```

Comment:
There is a push to use `done` instead of `next` in fastify (see https://github.com/fastify/fastify/pull/2786), so it might make sense to rename here for consistency as well.